### PR TITLE
MongoDB - loopback.getCurrentContext() return null

### LINKED ISSUE #809: loopback.getCurrentContext() return null after MongoDB call returns
A follow-up for #766.

I tested on https://github.com/strongloop/loopback-example-app

Step to create bug like this:
1. run npm install strongloop/loopback 
2. Add the code below into Customer.js
3. Add debug codes
4. Run app
6. Go to explorer page and make a request GET /Customers
7. Go to explorer page and Login to get AccessToken
8. Go to explorer page and set AccessToken to header in Explorer page
9. Go to explorer page and make a request GET /Customers
10. Check console log

My code:
```javascript
var loopback = require('loopback');

module.exports = function(Customer) {
  
  Customer.beforeRemote('**', function(remotingCtx, unused, next) {
    loopbackCtx = loopback.getCurrentContext();
    console.log("beforeRemote", loopbackCtx);
    loopbackCtx.modelData = 'aaaaaaaaaaaaaaa';
    next();
  });

  Customer.beforeSave =  function(next, customer) {
    loopbackCtx = loopback.getCurrentContext();
    console.log("beforeSave", loopbackCtx);
    next();
  };
};
```
and my log:
```plain
supervisor running without clustering (unsupervised)
Browse your REST API at http://localhost:3000/explorer
Web server listening at: http://localhost:3000/
beforeSave null
beforeSave null
beforeSave null
beforeSave null
beforeRemote { name: 'loopback',
  active: { accessToken: undefined },
  _set: [ null, { accessToken: undefined } ],
  id: 
   { create: [Function],
     flags: 15,
     before: [Function],
     after: [Function],
     error: [Function],
     uid: 1,
     data: null } }
beforeRemote null
```

### PR DESCRIPTION
**Pull request by projectxmaker**
loopback.getCurrentContext() return null after MongoDB call returns.

wrap up @Ichenay's patch https://github.com/lchenay/loopback/commit/f075c79 (see this comment for more details: https://github.com/strongloop/loopback/pull/337#issuecomment-61680577)

Fix #809

### CODE REVIEW

**Comment by bajtos** [comment]
Excuse my late reply, I was on holiday. Please agree which version should be moved forward. If it's the version from @doublemarked, then please open a new pull request.

Also change the name of the test to describe the scenario being tested (e.g. "preserves domain across async calls").

**Comment by doublemarked** [comment]
@projectxmaker  - let me know how you prefer to proceed

File: test/context.test.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    // remote method
+    TestModel.remoteMethod('test', {
+      accepts: {arg: 'inst', type: 'TestModel'},
+      returns: {root: true},
+      http: {path: '/test', verb: 'get'}
+    });
+
+    // after remote hook
+    TestModel.afterRemote('**', function(ctxx, inst, next){
+      ctxx.result.called = called;
```
Comment (projectxmaker) [inline test/context.test.js]: removed

File: test/context.test.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    app.model(TestModel, {dataSource: "db", public: true});
+
+    var emitterInOtherDomain = new EventEmitter();
+    Domain.create().add(emitterInOtherDomain);
+
+    runInOtherDomain = function(fn) {
+      emitterInOtherDomain.once('run', fn);
+    }
+
+    runnerInterval = setInterval(function() { emitterInOtherDomain.emit('run'); }, 10);
```
Comment (projectxmaker) [inline test/context.test.js]: I can't get what you described about "before/after hooks" here, could you give me more details ?

File: test/context.test.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
+var loopback = require('loopback');
+var request = require('supertest');
+var Domain = require('domain');
+var assert = require('assert');
+var EventEmitter = require("events").EventEmitter;
+
+describe('check loopback.getCurrentContext', function() {
```
Comment (projectxmaker) [inline test/context.test.js]: fixed

**Comment by projectxmaker** [comment]
@doublemarked let keep giving a hand in fixing the issue of the test file if you have time ^^

@bajtos I still need your help in fixing [this comment](https://github.com/strongloop/loopback/pull/885/files#r21896035)

**Comment by bajtos** [comment]
To speed the things up, I have made the final amendments myself and landed the patch via 989abf6:
 - reorganize the test code
 - give the test a descriptive name
 - fix remaining linter issues (e.g. missing EOL at EOF)
 - squash all commits into a single one, make the commit message descriptive

Thank you for the contribution!

@doublemarked @projectxmaker could you please verify that the bug is fixed, before I make a release? Just run `npm install strongloop/loopback` inside your project and then verify that your app works as expected.

**Comment by projectxmaker** [comment]
cool, I tested and it works properly ;)

**Comment by bajtos** [comment]
Thanks, I have released the fix as `loopback@2.8.8`.

**Comment by doublemarked** [comment]
Hey guys - looks great for me as well. Thank you for the release!

**Comment by STRML** [comment]
Please see https://github.com/strongloop/loopback/issues/1080 - domains are tricky.

**Comment by dongmai** [comment]
Hi @bajtos  ,

I'm facing this issue on the nearly latest version. Bellow is my `pacakge.json`

```
{
  "name": "test",
  "version": "1.0.0",
  "main": "server/server.js",
  "scripts": {
    "start": "node .",
    "pretest": "eslint .",
    "posttest": "nsp check"
  },
  "dependencies": {
    "async": "^2.0.0",
    "compression": "^1.0.3",
    "cors": "^2.5.2",
    "helmet": "^1.3.0",
    "loopback": "^2.22.0",
    "loopback-boot": "^2.6.5",
    "loopback-component-explorer": "^2.4.0",
    "loopback-connector-mongodb": "^1.15.2",
    "loopback-datasource-juggler": "^2.39.0",
    "loopback-testing": "^1.2.0",
    "serve-favicon": "^2.0.1"
  },
  "devDependencies": {
    "eslint": "^2.5.3",
    "nsp": "^2.1.0"
  },
  "repository": {
    "type": "",
    "url": ""
  },
  "license": "UNLICENSED",
  "description": "test"
}
```



I've used a middleware to save current logged user data into context but it (`getCurrentContext`) always returned `null` when working with MongoDB (memory db works just fine.) 

Bellow is my code, 

```
app.use(loopback.token()); 
app.use(loopback.context());

app.use(function (req, res, next) {
    if (!req.accessToken) return next();

    app.models.User.findById(req.accessToken.userId, function(err, user) {
        if (err)   return next(err);
        if (!user) return next(new Error('No user with this access token was found.'));
        res.locals.currentUser = user;

        var loopbackContext = loopback.getCurrentContext();
        console.log('SERVER CTX?' +loopbackContext);
        if (loopbackContext) {
            loopbackContext.set('currentUser', user);
        }else{
            throw new Error('Current context is null.');
        }
        next();
    });
});
```

Can you please have a look ? Have I done something wrong ?

Thanks.

**Comment by hash3r** [comment]
Have just the same problem with getCurrentContext. It returns null

**Pull request closed without merging**